vm/src/stdlib/array.rs

```diff
@@ -464,6 +467,40 @@ fn f64_try_into_from_object(vm: &VirtualMachine, obj: PyObjectRef) -> PyResult<f
     IntoPyFloat::try_from_object(vm, obj).map(|x| x.to_f64())
 }
 
+impl ArrayElement for WideChar {
+    fn try_into_from_object(vm: &VirtualMachine, obj: PyObjectRef) -> PyResult<Self> {
+        PyStrRef::try_from_object(vm, obj)?
+            .as_str()
+            .chars()
+            .exactly_one()
+            .map(|ch| Self(ch as _))
+            .map_err(|_| vm.new_type_error("array item must be unicode character".into()))
+    }
+
+    fn byteswap(self) -> Self {
+        Self(self.0.swap_bytes())
+    }
+}
+
+impl From<WideChar> for char {
+    fn from(ch: WideChar) -> Self {
+        // safe because every configuration of bytes for the types we support are valid
```

Comment:
I am not sure if it is actully safe. Unfortunately, our rust string implementation for python string has a few problems. One of them is this kind of thing.
Rust char only take valid unicode codepoints as its characters, while python is generally not. So if there isn't addiotional guarantee of validity.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2032ef87f8862e128d3f7150cb9d9365

Reply:
Thanks for carefully checking.

I tested with CPython, and here is the result:
```
>>> a = array.array('u', '\ud800')
>>> a[0]
'\ud800'
>>> print(a[0])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'utf-8' codec can't encode character '\ud800' in position 0: surrogates not allowed
```
Maybe we should `impl TryFrom<WideChar> for char` instead of `From<WideChar>`, and then we can throw an `UnicodeEncodeError` if the `u32` value is not a valid unicode point.

Reply:
I found another bug with string:

CPython:
```
>>> a = '\ud800'
>>> a[0]
'\ud800'
>>> print(a[0])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'utf-8' codec can't encode character '\ud800' in position 0: surrogates not allowed
```

RustPython
```
>>>>> a = '\ud800'
>>>>> a[0]
'�'
>>>>> print(a[0])
�
```

Reply:
`TryFrom<WideChar>` sounds like a good idea. We'd better to overhaul string implementation one day...